alf/utils/image_networks.py

```diff
@@ -0,0 +1,216 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import gin.tf
+import tensorflow as tf
+
+
+@gin.configurable
+class ImageEncodingNetwork(tf.keras.models.Model):
+    """
+    A general template class for create convolutional encoding networks.
+    """
+
+    def __init__(self,
+                 conv_layer_params,
+                 padding="valid",
+                 postprocess_fc_layer_params=None,
+                 activation_fn=tf.nn.relu,
+                 output_activation_fn=None,
+                 name="image_encoding_network"):
+        """
+        Initialize the layers for encoding an image into a latent vector.
+
+        Args:
+            conv_layer_params (list[tuple]): a non-empty list of elements
+                (num_filters, kernel_size, strides).
+            padding (str): "same" or "valid", see tf.keras.layers.Conv2DTranspose
+            postprocess_fc_layer_params (tuple[int]): a list of fc layer units.
+                These fc layers are used for postprocessing the conv output. If
+                None, then the output would preserve the shape of an image (will
+                not be flattened).
+            activation_fn (tf.nn.activation): activation for conv layers
+            output_activation_fn (tf.nn.activation): only used when there are
+                postprocessing FC layers
+            name (str): network name
+        """
+        super().__init__(name=name)
+
+        assert isinstance(conv_layer_params, list)
```

Comment:
This assertion seems to be inconsistent with the comment on support tuple as well ("conv_layer_params (list[tuple])").

Reply:
list[tuple] means "a list of tuples", not "list of tuple"